File: sdk/python/aegisrun/tool_call.py
"""Tool call execution"""

from typing import Dict, Any, Optional
from .client import AegisRunClient
from .types import Decision, PolicyAction


class ToolCall:
    """Represents a single tool invocation"""

    def __init__(
        self,
        client: AegisRunClient,
        run_id: str,
        step_id: str,
        seq_no: int,
        tool_name: str,
    ):
        self.client = client
        self.run_id = run_id
        self.step_id = step_id
        self.seq_no = seq_no
        self.tool_name = tool_name

        self.tool_call_id: Optional[str] = None
        self.decision: Optional[Decision] = None
        self.result: Optional[Any] = None
        self.error: Optional[str] = None

    def execute(
        self,
        args: Dict[str, Any],
        state_vector: Dict[str, Any],
        executor: str = "builtin",
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Any:
        """Execute the tool call"""
        response = self.client.execute_tool_call(
            run_id=self.run_id,
            step_id=self.step_id,
            tool_name=self.tool_name,
            args=args,
            state_vector=state_vector,
            metadata=metadata,
            executor=executor,
        )

        self.tool_call_id = response["tool_call_id"]

        # Parse decision as a struct: {action, policy_rule_id, reason, approval_id}
        raw_decision = response.get("decision", {})
        if isinstance(raw_decision, dict):
            self.decision = Decision(**raw_decision)
        else:
            self.decision = Decision(action=PolicyAction(raw_decision))

        self.result = response.get("result")
        self.error = response.get("error")

        # Handle blocked calls
        if self.decision and self.decision.action == PolicyAction.BLOCK:
            raise ToolCallBlockedError(
                f"Tool call blocked by policy: {self.decision.reason}",
                decision=self.decision,
            )

        if self.error:
            raise ToolCallExecutionError(self.error)

        return self.result


class ToolCallBlockedError(Exception):
    """Raised when a tool call is blocked by policy"""

    def __init__(self, message: str, decision: Decision):
        super().__init__(message)
        self.decision = decision


class ToolCallExecutionError(Exception):
    """Raised when tool execution fails"""
    pass

File: sdk/typescript/src/toolCall.ts
/**
 * Tool call execution
 */

import { AegisRunClient } from './client';
import { Decision } from './types';

export interface ToolCallConfig {
  client: AegisRunClient;
  runId: string;
  stepId: string;
  seqNo: number;
  toolName: string;
}

export class ToolCall {
  private client: AegisRunClient;
  private runId: string;
  private stepId: string;
  private seqNo: number;
  private toolName: string;

  public toolCallId?: string;
  public decision?: Decision;
  public result?: any;
  public error?: string;

  constructor(config: ToolCallConfig) {
    this.client = config.client;
    this.runId = config.runId;
    this.stepId = config.stepId;
    this.seqNo = config.seqNo;
    this.toolName = config.toolName;
  }

  async execute(params: {
    args: Record<string, any>;
    stateVector: Record<string, any>;
    executor?: string;
    metadata?: Record<string, any>;
  }): Promise<any> {
    const response = await this.client.executeToolCall({
      run_id: this.runId,
      step_id: this.stepId,
      tool_name: this.toolName,
      args: params.args,
      state_vector: params.stateVector,
      executor: params.executor,
      metadata: params.metadata,
    });

    this.toolCallId = response.tool_call_id;

    // Parse decision as a struct object
    if (response.decision) {
      this.decision = typeof response.decision === 'string'
        ? { action: response.decision as any, policy_rule_id: '', reason: '' }
        : response.decision;
    }

    if (response.result !== undefined) {
      this.result = response.result;
    }

    if (response.error) {
      this.error = response.error;
    }

    // Handle blocked calls
    if (this.decision && this.decision.action === 'block') {
      throw new ToolCallBlockedError(
        `Tool call blocked by policy`,
        this.decision
      );
    }

    if (this.error) {
      throw new ToolCallExecutionError(this.error);
    }

    return this.result;
  }
}

export class ToolCallBlockedError extends Error {
  constructor(message: string, public decision: Decision) {
    super(message);
    this.name = 'ToolCallBlockedError';
  }
}

export class ToolCallExecutionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ToolCallExecutionError';
  }
}
